internal/provider/resource_vpc_firewall_rules_v0.go

```diff
@@ -0,0 +1,724 @@
+		Timeouts:     m.Timeouts,
+		VPCID:        m.VPCID,
+		TimeCreated:  m.TimeCreated,
+		TimeModified: m.TimeModified,
+	}, nil
+}
+
+type vpcFirewallRulesResourceRuleModelV00 struct {
+	Action      types.String                                 `tfsdk:"action"`
+	Description types.String                                 `tfsdk:"description"`
+	Direction   types.String                                 `tfsdk:"direction"`
+	Filters     *vpcFirewallRulesResourceRuleFiltersModelV00 `tfsdk:"filters"`
+	Name        types.String                                 `tfsdk:"name"`
+	Priority    types.Int64                                  `tfsdk:"priority"`
+	Status      types.String                                 `tfsdk:"status"`
+	Targets     []vpcFirewallRulesResourceRuleTargetModelV00 `tfsdk:"targets"`
+
+	// Used to retrieve the timestamps from the API and populate the same fields
+	// within [vpcFirewallRulesResourceModel]. The `tfsdk:"-"` struct field tag is used
+	// to tell Terraform not to populate these values in the schema.
+	TimeCreated  types.String `tfsdk:"-"`
+	TimeModified types.String `tfsdk:"-"`
+}
+
+func (r vpcFirewallRulesResourceRuleModelV00) upgrade(ctx context.Context) (vpcFirewallRulesResourceRuleModelV01, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	targets := make([]vpcFirewallRulesResourceRuleTargetModelV01, len(r.Targets))
+	for i, t := range r.Targets {
+		targets[i] = vpcFirewallRulesResourceRuleTargetModelV01(t)
+	}
+
+	filters, filtersDiags := r.Filters.upgrade(ctx)
+	diags.Append(filtersDiags...)
+
+	if diags.HasError() {
+		return vpcFirewallRulesResourceRuleModelV01{}, diags
+	}
+
+	return vpcFirewallRulesResourceRuleModelV01{
+		Action:       r.Action,
+		Description:  r.Description,
+		Direction:    r.Direction,
+		Filters:      filters,
+		Name:         r.Name,
+		Priority:     r.Priority,
+		Status:       r.Status,
+		Targets:      targets,
+		TimeCreated:  r.TimeCreated,
+		TimeModified: r.TimeModified,
+	}, nil
+}
+
+type vpcFirewallRulesResourceRuleFiltersModelV00 struct {
+	Hosts     []vpcFirewallRuleHostFilterModelV00 `tfsdk:"hosts"`
+	Ports     types.Set                           `tfsdk:"ports"`
+	Protocols types.Set                           `tfsdk:"protocols"`
+}
+
+func (f *vpcFirewallRulesResourceRuleFiltersModelV00) upgrade(ctx context.Context) (*vpcFirewallRulesResourceRuleFiltersModelV01, diag.Diagnostics) {
+	if f == nil {
+		return nil, nil
+	}
+
+	hosts := make([]vpcFirewallRuleHostFilterModelV01, len(f.Hosts))
+	for i, h := range f.Hosts {
+		hosts[i] = vpcFirewallRuleHostFilterModelV01(h)
+	}
+
+	var protocols []types.String
+	diags := f.Protocols.ElementsAs(ctx, &protocols, false)
+	if diags.HasError() {
+		return nil, diags
+	}
+
+	upgradedProtocols := make([]vpcFirewallRuleProtocolFilterModelV01, len(protocols))
+	for i, p := range protocols {
+		upgradedProtocols[i] = vpcFirewallRuleProtocolFilterModelV01{
+			Type: types.StringValue(strings.ToLower(p.ValueString())),
+		}
+	}
```

Comment:
Most of the code here is copy-paste and boilerplate to ensure Terraform has a stable schema when loading older state files. This is where the schema upgrade actually happens.